src/vendor/vendor.c

```diff
@@ -125,22 +106,192 @@ int RDK_SetGroup(int group_id, kv_vector_t *params, unsigned *param_types, int *
 int WriteDMConfig(char *filename, char *mode, kv_vector_t *kvv, char *comment);
-void *rdk_malloc(size_t size);
-void rdk_free(void *ptr);
 
 #ifdef    INCLUDE_LCM_DATAMODEL
 #include "lcm_rbus_datamodel.c"
 #endif
+#define DEV_IP_DIA "Device.IP.Diagnostics.IPPing."
+
+// Array of valid input arguments
+static char *pingtest_input_args[] =
+{
+    "Host",
+};
+
+// Array of valid output arguments
+static char *pingtest_output_args[] =
+{
+    "Status",
+    "SuccessCount",
+    "FailureCount",
+};
+
+// input results of pingtest
+typedef struct
+{
+    int request_instance;
+}pingtest_input_cond_t;
+
+// Output results of pingtest
+typedef struct
+{
+    char Status[256];
+    char SuccessCount[256];
+    char FailureCount[256];
+} pingtest_output_res_t;
+
+/*********************************************************************//**
+**
+** PingThreadMain
+**
+** Called to log the status, SuccessCount and Failure Count of Device.IP.Diagnostics.IPPing() operation
+**
+** \param   params - pointer to input condition
+**
+** \return  USP_ERR_OK if successful
+**
+**************************************************************************/
+void *PingThreadMain(void* params)
+{
+    rbusValue_t values = NULL;
+    int rbus_err,count=4;
+    pingtest_input_cond_t *cond = (pingtest_input_cond_t *) params;
+    pingtest_output_res_t results;
+    pingtest_output_res_t *res = &results;
+    kv_vector_t *output_args;
+    int err = USP_ERR_OK;
+    char *err_msg = NULL, path[256];
+    memset(&results, 0, sizeof(results));
+    err = USP_SIGNAL_OperationStatus(cond->request_instance, "Active");
+    if (err != USP_ERR_OK)
+    {
+        USP_SNPRINTF(err_msg, sizeof(err_msg), "%s: USP_SIGNAL_OperationStatus() failed", __FUNCTION__);
+        goto exit;
+    }
+
+    do {
+        sleep(3);
+        rbus_err = rbus_get(bus_handle, "Device.IP.Diagnostics.IPPing.DiagnosticsState",&values);
+        if (rbus_err != RBUS_ERROR_SUCCESS)
+        {
+            USP_ERR_SetMessage("%s: rbus_get() failed (%d - %s)", __FUNCTION__, rbus_err, ToRbusErrString(rbus_err));
+            break;
+        }
+
+        strncpy(res->Status, rbusValue_ToString(values, NULL, 0), sizeof(res->Status) - 1);
+        if(strcmp(res->Status,"Complete")==0)
+        {
+            break;
+        }
+    } while (count--);
+    USP_LOG_Info("=========== IPPing Result ==========");
+    USP_LOG_Info("Status = %s", res->Status);
+    for (int j = 1; j < NUM_ELEM(pingtest_output_args); j++)
+    {
+        free(values);
+        sprintf(path, "%s%s", DEV_IP_DIA, pingtest_output_args[j]);
+        rbus_err = rbus_get(bus_handle, path, &values);
+        if (rbus_err != RBUS_ERROR_SUCCESS)
+        {
+            USP_ERR_SetMessage("%s: rbus_get() failed (%d - %s)", __FUNCTION__, rbus_err, ToRbusErrString(rbus_err));
+            goto exit;
+        }
+        const char* value_str = rbusValue_ToString(values, NULL, 0);
```

Comment:
The rbusValue_ToString allocates a buffer and returns to `value_str` must be released after the usage.
If `values` is holding the string data, we can use `rbusValue_GetString` which doesn't need explicit free.

Reply:
Hi @sundar2k ,
Released the values at the end of the for loop (rbusValue_Release(values);) at 227 line.
If releasing the value is immediately after the usage, then will update comment

Reply:
@sundar2k is saying that you also need to `free(value_str)` because `rbusValue_ToString` allocates a new buffer. If you instead use `rbusValue_GetString`, then you do not need to free `value_str`.

Reply:
okay @tmo-rshea3